addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/helper/WSHelper.java

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.bosesoundtouch.helper;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.HttpURLConnection;
+import java.net.ProtocolException;
+import java.net.URL;
+
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.openhab.binding.bosesoundtouch.handler.BoseSoundTouchHandler.BSTKeys;
+import org.openhab.binding.bosesoundtouch.handler.BoseSoundTouchHandler.KeyState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link WSHelper} class communicates with the REST API of the speaker.
+ *
+ * @author syracom - Initial contribution
+ */
+public class WSHelper implements WSHelperInterface {
+    private static final String KEY_PATH = "/key";
+    private Logger logger = LoggerFactory.getLogger(WSHelper.class);
+    private String deviceUrl;
+
+    public enum MESSAGE_TYPE {
+        POST,
+        GET
+    }
+
+    public WSHelper(String url) {
+        deviceUrl = url;
+    }
+
+    @Override
+    public synchronized String pressAndReleaseButtonOnSpeaker(BSTKeys keyIdentifier) {
+        return actionButtonOnSpeaker(keyIdentifier, KeyState.PRESS)
+                + actionButtonOnSpeaker(keyIdentifier, KeyState.RELEASE);
+    }
+
+    private String actionButtonOnSpeaker(BSTKeys keyIdentifier, KeyState keyState) {
+        String message = "<key state=\"" + keyState.getValue() + "\" sender=\"Gabbo\">" + keyIdentifier + "</key>";
+        String response = sendMessage(message, KEY_PATH, MESSAGE_TYPE.POST);
+        return checkForErrorFromSpeaker(response);
+    }
+
+    private String checkForErrorFromSpeaker(String response) {
+        if (response.startsWith("<error")) {
+            return "error";
+        }
+        return response;
+    }
+
+    @Override
+    public synchronized String setVolume(PercentType num) {
+        String message = "<volume>" + num + "</volume>";
+        return checkForErrorFromSpeaker(sendMessage(message, "/volume", MESSAGE_TYPE.POST));
+    }
+
+    private String sendMessage(String message, String key, MESSAGE_TYPE type) {
+        String url = deviceUrl + key;
+
+        BufferedReader in = null;
+        StringBuffer response = new StringBuffer();
```

Comment:
`StringBuilder` is faster than `StringBuffer` because it's not synchronized. So you should always use `StringBuilder`
 
And the thread safety which is delivered by `StringBuffer` is usually not enough because different append's can still be interleaved which still ruin your `StringBuffer`.

Just memorize to use the one which is the furthest in the Alphabet.